watchers/configmap_watcher.go

```diff
@@ -0,0 +1,60 @@
+package watcher
+
+import (
+	"context"
+	"os"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var logger = log.FromContext(context.TODO())
+
+func WatchForChanges(clientset *kubernetes.Clientset, namespace string, configMapName string, dataField string, dataValue string) {
+	logger.V(1).Info("watch for changes", "namespace", namespace, "configMap name", configMapName, "data field", dataField, "current value", dataValue)
+	mutex := &sync.Mutex{}
+	for {
+		watcher, err := clientset.CoreV1().ConfigMaps(namespace).Watch(context.TODO(),
+			metav1.SingleObject(metav1.ObjectMeta{Name: configMapName, Namespace: namespace}))
+		if err != nil {
+			logger.Error(err, "cannot create watcher", "namespace", namespace, "configMap name", configMapName)
+			os.Exit(1)
+		}
+		checkConfigMapChanges(watcher.ResultChan(), dataField, dataValue, mutex)
+	}
+}
+
+func checkConfigMapChanges(eventChannel <-chan watch.Event, dataField string, dataValue string, mutex *sync.Mutex) {
+	for {
+		event, open := <-eventChannel
+		if open {
+			switch event.Type {
+			case watch.Added:
+				fallthrough
+			case watch.Modified:
+				mutex.Lock()
+				defer mutex.Unlock()
```

Comment:
defer function is executed at the end of surrounding function; in this case it will (almost) never be executed, as the function returns only whe the eventChannel is closed and the infinite loop is broken.
One solution is to wrap whatever the whole case in an anonymous function.

Reply:
rethinking the need for this mutex: the main function is executed in a single thread.
i don't think there will be a concurrent access here.
wdyt of removing the mutex entirely ?

Reply:
I agree - we don't access or modify any shared state; the only reaction is `os.Exit(1)` if any.